tsl/src/continuous_aggs/common.c

```diff
@@ -0,0 +1,1379 @@
+static void caggtimebucketinfo_init(CAggTimebucketInfo *src, int32 hypertable_id,
+									Oid hypertable_oid, AttrNumber hypertable_partition_colno,
+									Oid hypertable_partition_coltype,
+									int64 hypertable_partition_col_interval,
+									int32 parent_mat_hypertable_id);
+static void caggtimebucket_validate(CAggTimebucketInfo *tbinfo, List *groupClause,
+									List *targetList);
+static bool cagg_agg_validate(Node *node, void *context);
+static bool cagg_query_supported(const Query *query, StringInfo hint, StringInfo detail,
+								 const bool finalized);
+static Oid cagg_get_boundary_converter_funcoid(Oid typoid);
+static FuncExpr *build_conversion_call(Oid type, FuncExpr *boundary);
+static FuncExpr *build_boundary_call(int32 ht_id, Oid type);
+static Const *cagg_boundary_make_lower_bound(Oid type);
+static Node *build_union_query_quals(int32 ht_id, Oid partcoltype, Oid opno, int varno,
+									 AttrNumber attno);
+static RangeTblEntry *makeRangeTblEntry(Query *subquery, const char *aliasname);
+
+static Const *
+check_time_bucket_argument(Node *arg, char *position)
+{
+	if (IsA(arg, NamedArgExpr))
+		arg = (Node *) castNode(NamedArgExpr, arg)->arg;
+
+	Node *expr = eval_const_expressions(NULL, arg);
+
+	if (!IsA(expr, Const))
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("only immutable expressions allowed in time bucket function"),
+				 errhint("Use an immutable expression as %s argument to the time bucket function.",
+						 position)));
+
+	return castNode(Const, expr);
+}
+
+/*
+ * Initialize caggtimebucket.
+ */
+static void
+caggtimebucketinfo_init(CAggTimebucketInfo *src, int32 hypertable_id, Oid hypertable_oid,
+						AttrNumber hypertable_partition_colno, Oid hypertable_partition_coltype,
+						int64 hypertable_partition_col_interval, int32 parent_mat_hypertable_id)
+{
+	src->htid = hypertable_id;
+	src->parent_mat_hypertable_id = parent_mat_hypertable_id;
+	src->htoid = hypertable_oid;
+	src->htpartcolno = hypertable_partition_colno;
+	src->htpartcoltype = hypertable_partition_coltype;
+	src->htpartcol_interval_len = hypertable_partition_col_interval;
+	src->bucket_width = 0;				 /* invalid value */
+	src->bucket_width_type = InvalidOid; /* invalid oid */
+	src->interval = NULL;				 /* not specified by default */
+	src->timezone = NULL;				 /* not specified by default */
+	TIMESTAMP_NOBEGIN(src->origin);		 /* origin is not specified by default */
+}
+
+/*
+ * Check if the supplied OID belongs to a valid bucket function
+ * for continuous aggregates.
+ */
+bool
+function_allowed_in_cagg_definition(Oid funcid)
+{
+	FuncInfo *finfo = ts_func_cache_get_bucketing_func(funcid);
+	if (finfo == NULL)
+		return false;
+
+	return finfo->allowed_in_cagg_definition;
+}
+
+/*
+ * Return Oid for a schema-qualified relation.
+ */
+Oid
+relation_oid(NameData schema, NameData name)
+{
+	return get_relname_relid(NameStr(name), get_namespace_oid(NameStr(schema), false));
+}
```

Comment:
The type `NameData` is a struct, and even though it is small, it does not seem necessary to copy it with each function call.

```suggestion
Oid
relation_oid(Name schema, Name name)
{
	return get_relname_relid(NameStr(*name), get_namespace_oid(NameStr(*schema), false));
}
```